internal/driver/controllerserver_helper.go

```diff
@@ -649,8 +692,31 @@ func (cs *ControllerServer) attachVolume(ctx context.Context, volumeID, linodeID
 	log.V(4).Info("Entering attachVolume()", "volume_id", volumeID, "node_id", linodeID)
 	defer log.V(4).Info("Exiting attachVolume()")
 
+	// Get the details of the volume to check for attachment possibilities
+	volume, err := cs.client.GetVolume(ctx, volumeID)
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to retrieve volume details: %v", err)
+	}
+
+	// Check if the volume is encrypted
+	isEncrypted := volume.Encryption == "enabled"
+	var linode *linodego.Instance
+
+	// If the volume is encrypted, get the details of the Linode instance and check region compatibility
+	if isEncrypted {
+		linode, err = cs.client.GetInstance(ctx, linodeID)
+		if err != nil {
+			return status.Errorf(codes.Internal, "failed to retrieve Linode instance details: %v", err)
+		}
+
+		// Ensure the encrypted volume and Linode are in the same region
+		if volume.Region != linode.Region {
+			return status.Errorf(codes.FailedPrecondition, "encrypted volume and Linode instance must be in the same region for attachment")
+		}
+	}
```

Comment:
This should be done inside getAndValidateVolume() helper. That way you don't have to make these 2 API calls here. We can reduce redundancy here.

We have the linode instance obj and volume obj available in getAndValidateVolume() plus logically makes sense to put this validation there. 

attachVolume() should only be for attaching logic not validation

Reply:
Yup, I've removed this logic as the getAndValidateVolume() is already checking for region mismatch. This isn't needed anymore.